### ISSUE TITLE
Astro's `prefetch` doesn't work in non-Chromium-based browsers

### ISSUE THREAD

**Issue by smastrom** [withastro/astro#10464]
### Astro Info

```block
Astro                    v4.5.5
Node                     v21.6.1
System                   macOS (arm64)
Package Manager          pnpm
Output                   server
Adapter                  @astrojs/cloudflare
Integrations             none
```

### If this issue only occurs in one browser, which browser is a problem?

Firefox, Safari

### Describe the Bug

According to the [docs](https://docs.astro.build/en/guides/prefetch/), Astro prefetch should work on any browser whether or not View Transitions are enabled (as it is not mentioned otherwise).

I virutally tried any combination under my `astro.config.mjs`: `prefetch: true`, `prefetch.prefetchAll: true`, etc. And also to manually append the `data-astro-prefetch` attribute as explained in the docs and enable/disable ViewTransitions. 

Under the "Network" tab of both Firefox 124 (latest) and Safari 17.4 (latest) there is no way I can see the incoming documents.

The only way to see them in that tab is to manually call `prefetch()` in a script tag using `{ with: 'fetch' }`, however, those pages are then requested again to the server once I click on the correspondent links.

On Chrome and Edge (and I suppose any other Chromium-based browser) everything works as per docs.

I am not sure if this expected, if it's a bug or if I (hopefully) misinterpreted the docs and actually doing something wrong. Thank you.

### What's the expected result?

To work on Firefox and Safari as it works on Chrome. 

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-rdn9rn?file=src%2Flayouts%2FLayout.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by bluwy** [comment]
I used Safari exclusively when implementing this, so it should work. Are you testing this in dev or prod? The prefetch behaviour also relies on a proper `Cache-Control` header set, so depending on your deployment platform, they may set a short Cache-Control so that it's cached on the second request. In dev, `Cache-Control` is usually not set so you might not see the benefits.

IIRC Chrome implements some unique heuristics, e.g. checking date of request and guessing a proper cache time if `Cache-Control` is not set. So it can sometime work for Chrome. https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#heuristic_caching

If you're unsure if the prefetching is actually done by the script, you can enable debug/verbose logging in the browser console to see the prefetch logs (only in dev).

**Comment by smastrom** [comment]
Hi @bluwy and thanks for your reply.

So, I've copied the StackBlitz repo linked above to [GitHub](https://github.com/smastrom/astro-issue) and deployed it to Cloudflare. The result doesn't change. Instead of Chrome I am using Edge but the behavior on Chrome is the same. Sorry for the UI in Italian but it seems I cannot change it easily.

### Edge / Chrome

If I refresh the _about_ page, you can see _contacts_ and _index_ fetched in the network tab:

![Screenshot 2024-03-18 alle 11 47 25](https://github.com/withastro/astro/assets/60471784/934a404b-30a0-46cb-b6a3-620a63289148)

Then by navigating to _contacts_ it's clear that it navigates to a prefetched page: 

![Screenshot 2024-03-18 alle 11 43 05](https://github.com/withastro/astro/assets/60471784/c7fd5bfa-2615-464a-901c-70178968b6d4)

### Safari

If I refresh the _about_ page that's what I see in the network tab (no prefetched pages):
 
![Screenshot 2024-03-18 alle 11 48 55](https://github.com/withastro/astro/assets/60471784/6fe2edeb-0b4c-4c90-8250-d35de277658a)

If I navigate to _contacts_, it's clear that that page has been downloaded as it took 216ms (_"scarica"_ means _"downloaded"_) and doesn't come from the prefetch cache:

![Screenshot 2024-03-18 alle 11 51 42](https://github.com/withastro/astro/assets/60471784/f1ed28f5-2a31-48fe-9846-d9fc691ab229)

On Firefox I have the same identical behavior of Safari.

I have published 4 Astro websites in the last few months and they are all affetcted by the same prefetch issue hence I guess it's not just my device or the dev environment.

Thank you again for your reply.

**Comment by bluwy** [comment]
Thanks! That's very helpful to understand the current situation. It looks like:

- For Safari, they [don't support](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/prefetch#browser_compatibility) `<link rel="prefetch"` at all :( We do use a normal `fetch()` in some cases and I wonder if we can fallback to that for Safari.
- For Firefox, it seems like it does prefetch, but it returns `NS_BINDING_ABORTED` for me. Searching online, it seems like because the headers didn't return `Expires` or `Cache-Control` so it decided to not prefetch it. It started streaming the HTML, but immediately abort it after the first chunk.
    <img width="854" alt="image" src="https://github.com/withastro/astro/assets/34116392/a27637db-ae47-4e04-9cd2-5605379ad58f">

I'm not exactly sure how to make these bullet-proof yet, but this is interesting to think about.

**Comment by smastrom** [comment]
Great, thanks for providing additional insights! If I may add another couple of things I noticed:

Calling `prefetch, { with: 'fetch'}` in a script tag is the only way to trigger the fetch on Firefox and Safari:

```html
<script>
   import { prefetch } from 'astro:prefetch'

   document.addEventListener('astro:page-load', () => {
      prefetch('/', { with: 'fetch' })
      prefetch('/about', { with: 'fetch' })
      prefetch('/contact', { with: 'fetch' })
   })
</script>
```

I can then see the pages downloaded in the network tab on both browsers:

<img width="1330" alt="Screenshot 2024-03-18 alle 18 21 24" src="https://github.com/withastro/astro/assets/60471784/212f612f-3206-47d1-a1f3-770bd492c5d4">

But once I try to navigate to one of them, they're downloaded again:

<img width="1328" alt="Screenshot 2024-03-18 alle 18 21 38" src="https://github.com/withastro/astro/assets/60471784/961b8a99-9f75-4de3-80bc-6bff1949013f">

I've added a [branch](https://github.com/smastrom/astro-issue/tree/programmatic-prefetch) to the repo linked above and deployed it [Cloudflare](https://programmatic-prefetch.astro-issue.pages.dev) just in case.

Cheers!

**Comment by matthewp** [comment]
@bluwy is this a bug?

**Comment by bluwy** [comment]
Yeah, it should work for Safari and Firefox. I think we need to figure out a support path for it, otherwise we need to document the caveats.

**Comment by V3RON** [comment]
@bluwy 

> For Safari, they [don't support](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/prefetch#browser_compatibility) <link rel="prefetch" at all :( We do use a normal fetch() in some cases and I wonder if we can fallback to that for Safari.

Sure we can!

`document.createElement('link').relList.supports('prefetch')` returns `true` in Chromium and `false` in Safari. 

I'll investigate it further and probably open a PR if we agree on this solution.

**Comment by bluwy** [comment]
@V3RON I was just working on this today and I noticed that 😄 I'll be sharing my findings here shortly and open a PR to restructure things, so a PR won't be needed for now, but thanks!
